text/0000-binary-type.md

```diff
@@ -0,0 +1,86 @@
+
+Last but not least, extending tremor with a binary type, something that JSON isn't capable of representing lays the groundwork for adding more powerful types to tremor. It serves as a case study of the cost/complexity of this and will help simplifying the task for later additions.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The `binary` type in tremor comes along with additions to the tremor-script syntax, the type system, codecs, the standard library and a specification of the serialization behavior. We will discuss each of those 
+
+## The `binary` type
+The easiest way to think about it is that the `binary` type represents an array of bytes.
+
+## Serialization
+
+This is codec dependant and not always symmetric, can not represent binary data and is a general-purpose codec it will default to base64 encoding the data as a string (JSON for example). This is an asymmetric change as we will not by default decode a base64 string as binary given without context we can't determine if this is the desired behavior or not, this can however be done using tremor-scripts functions.
+
+## Syntax
+
+We use the (erlang bit syntax)[https://erlang.org/doc/programming_examples/bit_syntax.html] as an inspiration. Tremor script gains support for binary semi-literals using the form of `<< expr1, expr2 >>` where `expr` has to return a 1-byte number. More elaborate forms, such as bitwise size and different types are aimed to be added later and follow the erlang bit syntax.
+
+Binary matching will likely follow the same logical pattern as other matching syntaxes, and aim to provide at least erlang style dissection of binary data.
+
+## stdlib
+
+This RFC introduces the `base64` module for encoding and decoding `binary` data as base64 strings. It also adds the `binary` module with basic functions such as `len`, as well as conversion functions to and from arrays of numbers.
+
+The `string` module gains functions to convert a string to a `binary` as well as a function that will take a `binary` and convert it to a utf8 `string` in a potentially lossy fashion for invalid utf8 data.
+
+The `type` module will gain the `is_binary` function as well as `type::as_string` now returning `"binary"` when appropriate.
+
+## codecs
+
+This RFC introduces the `binary` codec that passes bytes unmodified as binary data.
+
+The `json` codec is changed to serialize `binary` data as a base64 encoded string, however it will not automatically decode base64 strings as binary.
+
+The `msgpack` codec should encode and decode binary as `bin` types.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `binary` type is a `Vec<u8>` that serializes as base64 encoded string in simd-json and as `bytes`.
```

Comment:
What is `bytes` here?

Reply:
That was a half finished sentence, `bytes` is a encoder in serde :)